sd/lb/retry.go

```diff
@@ -48,7 +68,18 @@ func Retry(max int, timeout time.Duration, b Balancer) endpoint.Endpoint {
 			case response := <-responses:
 				return response, nil
 			case err := <-errs:
-				a = append(a, err.Error())
+				cont, cbErr := cb(i, err)
+				if !cont {
+					if cbErr == nil {
+						return nil, fmt.Errorf("retry attempts exceeded (%s)", strings.Join(a, "; "))
+					}
+					return nil, cbErr
+				}
+				currentErr := err.Error()
+				if cbErr != nil {
+					currentErr = cbErr.Error()
+				}
+				a = append(a, currentErr)
```

Comment:
This is super messy, especially regarding the final error returned to the caller, which at this point I think deserves to be lifted into its own type, with a slice of underlying errors as a member. But, I think a lot of the messiness may go away if you're able to implement all of this logic in the callback as I wanted above.